aten/src/ATen/SavedTensorHooks.cpp

```diff
@@ -59,25 +65,19 @@ void SavedTensorDefaultHooks::push_hooks(PyObject* pack_hook, PyObject* unpack_h
   tls.stack.emplace(pack_hook, unpack_hook);
 }
 
-void SavedTensorDefaultHooks::pop_hooks() {
+std::pair<PyObject*, PyObject*> SavedTensorDefaultHooks::pop_hooks() {
   // Reference counting is handled by the caller of `pop_hooks`
   TORCH_INTERNAL_ASSERT(is_initialized && !tls.stack.empty());
+  std::pair<PyObject*, PyObject*> hooks = tls.stack.top();
   tls.stack.pop();
+  return hooks;
 }
 
 std::pair<PyObject*, PyObject*> SavedTensorDefaultHooks::get_hooks() {
-  if (!is_initialized || tls.stack.empty()) {
+  if (!is_initialized || tls.stack.empty() || tls.is_tracing) {
```

Comment:
maybe worth a comment about `disable_saved_tensors_hooks` and why we need this, instead of registering a no-op saved tensor hook?

cc @albanD @soulitzer

Reply:
Could the no-op saved tensor hook still work? If the concern is to avoid triggering the error from disable_saved_tensor_hooks, we could just not push the no-op hook in the case where saved tensor hooks are already disabled, because being in a disable_saved_tensor_hooks ctx means we are guaranteed that there are no saved tensor hooks running.

Btw to clarify, this PR is mainly to handle the case where there are saved tensor hooks outside the compiled region right?

Reply:
> we could just not push the no-op hook in the case where saved tensor hooks are already disabled

I added another test case `test_disabling_unpack_hooks_within_compiled_region` that would fail with pushing no-op hooks

This PR is to make saved tensor hooks fire exactly once in torch.compile